tools/deviceGen/output/traits_defaults.hlsl

```diff
@@ -0,0 +1,382 @@
+NBL_GENERATE_GET_OR_DEFAULT(maxMemoryAllocationCount, uint32_t, 4096);
+NBL_GENERATE_GET_OR_DEFAULT(maxSamplerAllocationCount, uint32_t, 4000);
+NBL_GENERATE_GET_OR_DEFAULT(bufferImageGranularity, uint32_t, 65536);
+NBL_GENERATE_GET_OR_DEFAULT(maxPerStageDescriptorSamplers, uint32_t, 16);
+NBL_GENERATE_GET_OR_DEFAULT(maxPerStageDescriptorUBOs, uint32_t, 15);
+NBL_GENERATE_GET_OR_DEFAULT(maxPerStageDescriptorSSBOs, uint32_t, 31);
+NBL_GENERATE_GET_OR_DEFAULT(maxPerStageDescriptorImages, uint32_t, 96);
+NBL_GENERATE_GET_OR_DEFAULT(maxPerStageDescriptorStorageImages, uint32_t, 8);
+NBL_GENERATE_GET_OR_DEFAULT(maxPerStageDescriptorInputAttachments, uint32_t, 7);
+NBL_GENERATE_GET_OR_DEFAULT(maxPerStageResources, uint32_t, 127);
+NBL_GENERATE_GET_OR_DEFAULT(maxDescriptorSetSamplers, uint32_t, 80);
+NBL_GENERATE_GET_OR_DEFAULT(maxDescriptorSetUBOs, uint32_t, 90);
+NBL_GENERATE_GET_OR_DEFAULT(maxDescriptorSetDynamicOffsetUBOs, uint32_t, 8);
+NBL_GENERATE_GET_OR_DEFAULT(maxDescriptorSetSSBOs, uint32_t, 155);
+NBL_GENERATE_GET_OR_DEFAULT(maxDescriptorSetDynamicOffsetSSBOs, uint32_t, 8);
+NBL_GENERATE_GET_OR_DEFAULT(maxDescriptorSetImages, uint32_t, 480);
+NBL_GENERATE_GET_OR_DEFAULT(maxDescriptorSetStorageImages, uint32_t, 40);
+NBL_GENERATE_GET_OR_DEFAULT(maxDescriptorSetInputAttachments, uint32_t, 7);
+NBL_GENERATE_GET_OR_DEFAULT(maxVertexOutputComponents, uint16_t, 124);
+NBL_GENERATE_GET_OR_DEFAULT(maxTessellationGenerationLevel, uint16_t, 0);
+NBL_GENERATE_GET_OR_DEFAULT(maxTessellationPatchSize, uint16_t, 0);
+NBL_GENERATE_GET_OR_DEFAULT(maxTessellationControlPerVertexInputComponents, uint16_t, 0);
+NBL_GENERATE_GET_OR_DEFAULT(maxTessellationControlPerVertexOutputComponents, uint16_t, 0);
+NBL_GENERATE_GET_OR_DEFAULT(maxTessellationControlPerPatchOutputComponents, uint16_t, 0);
+NBL_GENERATE_GET_OR_DEFAULT(maxTessellationControlTotalOutputComponents, uint16_t, 0);
+NBL_GENERATE_GET_OR_DEFAULT(maxTessellationEvaluationInputComponents, uint16_t, 0);
+NBL_GENERATE_GET_OR_DEFAULT(maxTessellationEvaluationOutputComponents, uint16_t, 0);
+NBL_GENERATE_GET_OR_DEFAULT(maxGeometryShaderInvocations, uint16_t, 0);
+NBL_GENERATE_GET_OR_DEFAULT(maxGeometryInputComponents, uint16_t, 0);
+NBL_GENERATE_GET_OR_DEFAULT(maxGeometryOutputComponents, uint16_t, 0);
+NBL_GENERATE_GET_OR_DEFAULT(maxGeometryOutputVertices, uint16_t, 0);
+NBL_GENERATE_GET_OR_DEFAULT(maxGeometryTotalOutputComponents, uint16_t, 0);
+NBL_GENERATE_GET_OR_DEFAULT(maxFragmentInputComponents, uint32_t, 116);
+NBL_GENERATE_GET_OR_DEFAULT(maxFragmentOutputAttachments, uint32_t, 8);
+NBL_GENERATE_GET_OR_DEFAULT(maxFragmentDualSrcAttachments, uint32_t, 1);
+NBL_GENERATE_GET_OR_DEFAULT(maxFragmentCombinedOutputResources, uint32_t, 16);
+NBL_GENERATE_GET_OR_DEFAULT(maxComputeSharedMemorySize, uint32_t, 32768);
+NBL_GENERATE_GET_OR_DEFAULT(maxComputeWorkGroupCount[3], uint32_t, {MinMaxWorkgroupCount,MinMaxWorkgroupCount,MinMaxWorkgroupCount});
```

Comment:
you need a different way to handle array/vector constants because `NBL_GENERATE_GET_OR_DEFAULT` is implemented via inheritance from an `integral_constant`

Reply:
[it will detect the member](https://godbolt.nsc.devsh.eu/#g:!((g:!((g:!((h:output,i:(compilerName:'NSC+(Release)',editorid:1,fontScale:14,fontUsePx:'0',j:1,wrap:'1'),l:'5',n:'0',o:'Output+of+NSC+(Release)+(Compiler+%231)',t:'0'),(h:compiler,i:(compiler:nsc_release_upstream,filters:(b:'0',binary:'1',binaryObject:'1',commentOnly:'0',debugCalls:'1',demangle:'0',directives:'0',execute:'1',intel:'0',libraryCode:'0',trim:'1'),flagsViewOpen:'1',fontScale:14,fontUsePx:'0',j:1,lang:hlsl,libs:!(),options:'-T+cs_6_7+-E+main',overrides:!(),selection:(endColumn:1,endLineNumber:1,positionColumn:1,positionLineNumber:1,selectionStartColumn:1,selectionStartLineNumber:1,startColumn:1,startLineNumber:1),source:1),l:'5',n:'0',o:'+NSC+(Release)+(Editor+%231)',t:'0')),k:50,l:'4',n:'0',o:'',s:1,t:'0'),(g:!((h:codeEditor,i:(filename:'1',fontScale:14,fontUsePx:'0',j:1,lang:hlsl,selection:(endColumn:14,endLineNumber:23,positionColumn:14,positionLineNumber:23,selectionStartColumn:14,selectionStartLineNumber:23,startColumn:14,startLineNumber:23),source:'%23include+%22nbl/builtin/hlsl/device_capabilities_traits.hlsl%22%0A%0ANBL_GENERATE_MEMBER_TESTER(maxComputeWorkGroupCount)%3B%0A%0A%0A%23define+NBL_GENERATE_GET_OR_DEFAULT(field,+ty,+default)+%5C%0Atemplate%3Ctypename+S,+bool+%3D+has_member_%23%23field%3CS%3E::value%3E+struct+get_or_default_%23%23field+:+integral_constant%3Cty,S::field%3E+%7B%7D%3B+%5C%0Atemplate%3Ctypename+S%3E+struct+get_or_default_%23%23field%3CS,false%3E+:+integral_constant%3Cty,default%3E+%7B%7D%3B%0A%0Anamespace+nbl%0A%7B%0Anamespace+hlsl%0A%7B%0Anamespace+impl%0A%7B%0ANBL_GENERATE_GET_OR_DEFAULT(maxComputeWorkGroupCount,bool,false)%3B%0A%7D%0A%7D%0A%7D%0A%0Astruct+Caps%0A%7B%0A++++uint32_t3+maxComputeWorkGroupCount%3B%0A%7D%3B%0A%0A%5Bnumthreads(1,1,1)%5D%0Avoid+main()%0A%7B%0A++++vk::RawBufferStore%3Cbool%3E(0xdeadbeefull,nbl::hlsl::has_member_maxComputeWorkGroupCount%3CCaps%3E::value)%3B%0A%7D'),l:'5',n:'0',o:'HLSL+source+%231',t:'0')),header:(),k:50,l:'4',n:'0',o:'',s:0,t:'0')),l:'2',n:'0',o:'',t:'0')),version:4)

Reply:
[the `integral_constant` cannot be made with a non-primitive type such as `uint32_t3` or `uint32_t3`](https://godbolt.nsc.devsh.eu/#g:!((g:!((g:!((h:output,i:(compilerName:'NSC+(Release)',editorid:1,fontScale:14,fontUsePx:'0',j:1,wrap:'1'),l:'5',n:'0',o:'Output+of+NSC+(Release)+(Compiler+%231)',t:'0'),(h:compiler,i:(compiler:nsc_release_upstream,filters:(b:'0',binary:'1',binaryObject:'1',commentOnly:'0',debugCalls:'1',demangle:'0',directives:'0',execute:'1',intel:'0',libraryCode:'0',trim:'1'),flagsViewOpen:'1',fontScale:14,fontUsePx:'0',j:1,lang:hlsl,libs:!(),options:'-T+cs_6_7+-E+main',overrides:!(),selection:(endColumn:1,endLineNumber:1,positionColumn:1,positionLineNumber:1,selectionStartColumn:1,selectionStartLineNumber:1,startColumn:1,startLineNumber:1),source:1),l:'5',n:'0',o:'+NSC+(Release)+(Editor+%231)',t:'0')),k:50,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:codeEditor,i:(filename:'1',fontScale:14,fontUsePx:'0',j:1,lang:hlsl,selection:(endColumn:87,endLineNumber:13,positionColumn:87,positionLineNumber:13,selectionStartColumn:87,selectionStartLineNumber:13,startColumn:87,startLineNumber:13),source:'%23include+%22nbl/builtin/hlsl/device_capabilities_traits.hlsl%22%0A%0ANBL_GENERATE_MEMBER_TESTER(maxComputeWorkGroupCount)%3B%0Astruct+Caps%0A%7B%0A++++NBL_CONSTEXPR_STATIC_INLINE+uint32_t3+maxComputeWorkGroupCount%3B%0A%7D%3B%0ANBL_CONSTEXPR_STATIC_INLINE+uint32_t3+Caps::maxComputeWorkGroupCount+%3D+%7B16,16,1%7D%3B%0A%0Ausing+test00_t+%3D+nbl::hlsl::integral_constant%3Cuint32_t%5B3%5D,%7B69,1,2%7D%3E%3B%0Ausing+test01_t+%3D+nbl::hlsl::integral_constant%3Cuint32_t%5B3%5D,Caps::maxComputeWorkGroupCount%3E%3B%0Ausing+test10_t+%3D+nbl::hlsl::integral_constant%3Cuint32_t3,%7B69,1,2%7D%3E%3B%0Ausing+test11_t+%3D+nbl::hlsl::integral_constant%3Cuint32_t3,Caps::maxComputeWorkGroupCount%3E%3B%0A%0A%5Bnumthreads(1,1,1)%5D%0Avoid+main()%0A%7B%0A%7D'),l:'5',n:'0',o:'HLSL+source+%231',t:'0')),header:(),k:50,l:'4',n:'0',o:'',s:0,t:'0')),l:'2',n:'0',o:'',t:'0')),version:4)

You need to break down vectors/arrays into multiple members with `X`, `Y`, `Z` suffices

Reply:
but obviously do no changes to the C++ structs